src/main/java/org/prebid/server/bidder/adnuntius/AdnuntiusBidder.java

```diff
@@ -241,37 +260,47 @@ private MultiMap getHeaders(Device device) {
         try {
             final String body = httpCall.getResponse().getBody();
-            final AdnuntiusResponse bidResponse = mapper.decodeValue(body, AdnuntiusResponse.class);
-            return Result.withValues(extractBids(bidRequest, bidResponse));
+            final AdnuntiusResponse adnuntiusResponse = mapper.decodeValue(body, AdnuntiusResponse.class);
+            return Result.withValues(extractBids(bidRequest, adnuntiusResponse));
         } catch (DecodeException | PreBidException e) {
             return Result.withError(BidderError.badServerResponse(e.getMessage()));
         }
     }
 
-    private static List<BidderBid> extractBids(BidRequest bidRequest, AdnuntiusResponse bidResponse) {
-        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getAdsUnits())) {
+    private List<BidderBid> extractBids(BidRequest bidRequest, AdnuntiusResponse adnuntiusResponse) {
+        if (adnuntiusResponse == null || CollectionUtils.isEmpty(adnuntiusResponse.getAdsUnits())) {
             return Collections.emptyList();
         }
 
-        final List<AdnuntiusAdsUnit> adsUnits = bidResponse.getAdsUnits();
+        final List<AdnuntiusAdsUnit> adsUnits = adnuntiusResponse.getAdsUnits();
         final List<Imp> imps = bidRequest.getImp();
         if (adsUnits.size() > imps.size()) {
             throw new PreBidException("Impressions count is less then ads units count.");
         }
 
-        final List<AdsUnitWithImpId> validAdsUnitToImpId = IntStream.range(0, adsUnits.size())
-                .mapToObj(i -> AdsUnitWithImpId.of(adsUnits.get(i), imps.get(i).getId()))
+        final List<AdsUnitWithImpId> validAdsUnitToImp = IntStream.range(0, adsUnits.size())
+                .mapToObj(i -> AdsUnitWithImpId.of(adsUnits.get(i),
+                        imps.get(i),
+                        parseImpExt(imps.get(i))))
                 .filter(adsUnitWithImpId -> validateAdsUnit(adsUnitWithImpId.getAdsUnit()))
                 .toList();
 
-        if (validAdsUnitToImpId.isEmpty()) {
+        if (validAdsUnitToImp.isEmpty()) {
             return Collections.emptyList();
         }
 
-        final String currency = extractCurrency(validAdsUnitToImpId);
-        return validAdsUnitToImpId.stream()
-                .map(adsUnitWithImpId -> makeBid(adsUnitWithImpId.getAdsUnit(), adsUnitWithImpId.getImpId(), currency))
-                .toList();
+        final String currency = extractCurrency(validAdsUnitToImp);
+
+        final List<BidderBid> bidderBids = new ArrayList<>(validAdsUnitToImp.stream()
+                .map(adsUnitWithImpId -> makeGeneralBid(adsUnitWithImpId, currency))
+                .toList());
+
+        bidderBids.addAll(validAdsUnitToImp.stream()
+                .filter(adsUnitWithImpId -> CollectionUtils.isNotEmpty(adsUnitWithImpId.getAdsUnit().getDeals()))
+                .map(adsUnitWithImpId -> makeDealsBid(adsUnitWithImpId, currency))
+                .toList());
+
+        return bidderBids;
```

Comment:
Can be simplified to:
```java
final String currency = extractCurrency(validAdsUnitToImp);
final Stream<BidderBid> generalBids = validAdsUnitToImp.stream()
                .map(adsUnitWithImpId -> makeGeneralBid(adsUnitWithImpId, currency));
final Stream<BidderBid> dealBids = validAdsUnitToImp.stream()
                .filter(adsUnitWithImpId -> CollectionUtils.isNotEmpty(adsUnitWithImpId.getAdsUnit().getDeals()))
                .map(adsUnitWithImpId -> makeDealsBid(adsUnitWithImpId, currency))
                .filter(Objects::nonNull);

return Stream.concat(generalBids, dealBids).toList();
```

No need to create temporary lists every time and put pressure on GC.